rfd/xxxx-config-resource-docs.md

```diff
@@ -0,0 +1,554 @@
+reference is particularly lagging, with only five resources documented at the
+[time of
+writing](https://github.com/gravitational/teleport/blob/58f20f0c3fcc9bb281528915125b95dfaee6cec5/docs/pages/reference/resources.mdx).
+
+## Details
+
+The generator reads a [configuration](#configuration) that specifies struct
+types to generate references from. After parsing Go source files, the generator
+uses the resulting AST information to [populate a template](#the-output). 
+
+To do so, it builds [mappings](#working-with-parsed-go-source) that it can use
+to track type declarations in source packages. Using these mappings, it
+[converts](#converting-structs-to-resources) struct types to template data. To
+document the types of struct fields, the generator writes [additional
+entries](#processing-field-types) to the reference template as appropriate.
+
+### Configuration
+
+The generator will have access to a mapping of package paths to lists of named
+struct types within each package. It will use this mapping to identify the
+struct types to use for generating the resource reference. 
+
+Since struct definitions for `tctl` resources include field types that also
+require their own entries in the resource reference (e.g., `types.Metadata`),
+the configuration only names the _root_ types for the generator to examine.
+
+We don't have a formal naming convention for the root types of each resource,
+and may not want to list *all* resources (e.g., there may be internal resources
+to keep hidden), so it would make sense to list resources explicitly. The CRD
+generator for the Teleport Kubernetes Operator [uses an explicit list of
+types](https://github.com/gravitational/teleport/blob/c9b0a601ab781e1bd34dc3c44a94e6e85e3e64ae/integrations/operator/crdgen/main.go#L96-L113),
+and the generator I'm proposing will also include package names so the generator
+loads the correct source packages and their dependencies.
+
+The program could read the mapping from a YAML configuration file or declare it
+directly as a Go `map`.
+
+Here is an example in YAML:
+
+```yaml
+github.com/gravitational/teleport/api/types
+- ProvisionTokenV2
+- AuthPreferenceV2
+- AccessRequestV3
+
+github.com/gravitational/teleport/api/loginrulev1
+- LoginRule
+```
+
+#### Alternative: Working from protobuf message definitions
+
+We could write a `protoc` plugin that generates the configuration resource
+reference from our protobuf message definitions. This is how we generate our
+Terraform provider source in `github.com/gravitational/protoc-gen-terraform`.
+
+However, we can't guarantee that every configuration resource available to
+`tctl` will correspond to a protobuf message definition. Since we generate Go
+source files based on our `.proto` files, and Teleport needs to unmarshal all
+configuration resources into structs from YAML, it's a safe bet that all
+configuration resources will have a corresponding Go struct.
```

Comment:
+1 for parsing Go sources too. It's pretty unlikely a resource won't have a Go representation, plus it doesn't tie us to necessarily to protos (or a particular proto generator).